<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="ipynb_website:version" content="0.9.7" />
<meta name="viewport" content="width=device-width, initial-scale=1" />

<link rel="stylesheet" type="text/css" href="../css/jt.css">

<link rel="stylesheet" type="text/css" href="../css/toc2.css">

<link href="../site_libs/jqueryui-1.11.4/jquery-ui.css">
<link rel="stylesheet" href="../site_libs/bootstrap-3.3.5/css/cosmo.min.css" rel="stylesheet" />
<link rel="stylesheet" href="../site_libs/font-awesome-4.5.0/css/font-awesome.min.css" rel="stylesheet" />
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jquery/1.9.1/jquery.min.js"></script>
<script type="text/javascript" src="https://ajax.googleapis.com/ajax/libs/jqueryui/1.9.1/jquery-ui.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/js/bootstrap.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/html5shiv.min.js"></script>
<script src="../site_libs/bootstrap-3.3.5/shim/respond.min.js"></script>

<link rel="stylesheet"
      href="../site_libs/highlightjs/null.min.css"
      type="text/css" />

<script src="../site_libs/highlightjs/highlight.pack.js"></script>
<script>hljs.initHighlightingOnLoad();</script>
<script type="text/javascript">
if (window.hljs && document.readyState && document.readyState === "complete") {
   window.setTimeout(function() {
      hljs.initHighlighting();
   }, 0);
}
</script>

<script src="../js/doc_toc.js"></script>
<script src="../js/docs.js"></script>

<script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.2/MathJax.js?config=TeX-MML-AM_CHTML"></script>
<script>
    MathJax.Hub.Config({
        extensions: ["tex2jax.js"],
        jax: ["input/TeX", "output/HTML-CSS"],
        tex2jax: {
        inlineMath: [ ['$','$'], ["\\(","\\)"] ],
        displayMath: [ ['$$','$$'], ["\\[","\\]"] ],
        processEscapes: true
        },
        "HTML-CSS": {
            preferredFont: "TeX",
            availableFonts: ["TeX"],
            styles: {
                scale: 110,
                ".MathJax_Display": {
                    "font-size": "110%",
                }
            }
        }
    });
</script>
<script>
function filterDataFrame(id) {
    var input = document.getElementById("search_" + id);
    var filter = input.value.toUpperCase();
    var table = document.getElementById("dataframe_" + id);
    var tr = table.getElementsByTagName("tr");
    // Loop through all table rows, and hide those who don't match the search query
    for (var i = 1; i < tr.length; i++) {
        for (var j = 0; j < tr[i].cells.length; ++j) {
            var matched = false;
            if (tr[i].cells[j].innerHTML.toUpperCase().indexOf(filter) != -1) {
                tr[i].style.display = "";
                matched = true
                break;
            }
            if (!matched)
                tr[i].style.display = "none";
        }
    }
}
function sortDataFrame(id, n, dtype) {
    var table = document.getElementById("dataframe_" + id);
    var tb = table.tBodies[0]; // use `<tbody>` to ignore `<thead>` and `<tfoot>` rows
    var tr = Array.prototype.slice.call(tb.rows, 0); // put rows into array
    if (dtype === 'numeric') {
        var fn = function(a, b) { 
            return parseFloat(a.cells[n].textContent) <= parseFloat(b.cells[n].textContent) ? -1 : 1;
        }
    } else {
        var fn = function(a, b) {
            var c = a.cells[n].textContent.trim().localeCompare(b.cells[n].textContent.trim()); 
            return c > 0 ? 1 : (c < 0 ? -1 : 0) }
    }
    var isSorted = function(array, fn) {
        if (array.length < 2)
            return 1;
        var direction = fn(array[0], array[1]); 
        for (var i = 1; i < array.length - 1; ++i) {
            var d = fn(array[i], array[i+1]);
            if (d == 0)
                continue;
            else if (direction == 0)
                direction = d;
            else if (direction != d)
                return 0;
            }
        return direction;
    }
    var sorted = isSorted(tr, fn);
    if (sorted == 1 || sorted == -1) {
        // if sorted already, reverse it
        for(var i = tr.length - 1; i >= 0; --i)
            tb.appendChild(tr[i]); // append each row in order
    } else {
        tr = tr.sort(fn);
        for(var i = 0; i < tr.length; ++i)
            tb.appendChild(tr[i]); // append each row in order
    }
}
</script>

<script>
$( document ).ready(function(){
            var cfg={'threshold':3,     // depth of toc (number of levels)
             'number_sections': false,
             'toc_cell': false,          // useless here
             'toc_window_display': true, // display the toc window
             "toc_section_display": "block", // display toc contents in the window
             'sideBar':true,       // sidebar or floating window
             'navigate_menu':false       // navigation menu (only in liveNotebook -- do not change)
            }
            var st={};                  // some variables used in the script
            st.rendering_toc_cell = false;
            st.config_loaded = false;
            st.extension_initialized=false;
            st.nbcontainer_marginleft = $('#notebook-container').css('margin-left')
            st.nbcontainer_marginright = $('#notebook-container').css('margin-right')
            st.nbcontainer_width = $('#notebook-container').css('width')
            st.oldTocHeight = undefined
            st.cell_toc = undefined;
            st.toc_index=0;
            // fire the main function with these parameters
            table_of_contents(cfg, st);
            var file=analysisDict[$("h1:first").attr("id")];
            $("#toc-level0 a").css("color","#126dce");
            $('a[href="#'+$("h1:first").attr("id")+'"]').hide()
            var docs=analysisArray;
            var docs_map=analysisArrayMap;
            var pos=analysisArray.indexOf(file);
            for (var a=pos;a>=0;a--){
                  $('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>').insertBefore("#toc-level0 li:eq(0)");
            }
            $('a[href="'+file+'.html'+'"]').css("color","#126dce");
            for (var a=pos+1;a<docs.length;a++){
                  $(".toc #toc-level0").append('<li><a href="'+docs[a]+'.html"><font color="#073642"><b>'+docs_map[docs[a]].replace(/_/g," ")+'</b></font></a></li>');
            }
            // $("#toc-header").hide(); // comment out because it prevents search bar from displaying
    });
</script>

<script>
// manage active state of menu based on current page
$(document).ready(function () {
  // active menu anchor
  href = window.location.pathname
  href = href.substr(href.lastIndexOf('/') + 1)
  if (href === "")
    href = "index.html";
  var menuAnchor = $('a[href="' + href + '"]');
  // mark it active
  menuAnchor.parent().addClass('active');
  // if it's got a parent navbar menu mark it active as well
  menuAnchor.closest('li.dropdown').addClass('active');
});
</script>
<div class="container-fluid main-container">
<!-- tabsets -->
<script src="../site_libs/navigation-1.1/tabsets.js"></script>
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>



<title>RV-NPL project</title>

<style type = "text/css">
body {
  font-family: "Droid Sans";
  padding-top: 66px;
  padding-bottom: 40px;
}
</style>
</head>

<body>
<div tabindex="-1" id="notebook" class="border-box-sizing">
<div class="container" id="notebook-container">

<!-- code folding -->

<div class="navbar navbar-default  navbar-fixed-top" role="navigation">
  <div class="container">
    <div class="navbar-header">
      <button type="button" class="navbar-toggle collapsed" data-toggle="collapse" data-target="#navbar">
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
        <span class="icon-bar"></span>
      </button>
      <a class="navbar-brand" href="../index.html">RV-NPL project</a>
    </div>
    <div id="navbar" class="navbar-collapse collapse">
      <ul class="nav navbar-nav">
        
<li>
  <a href="../index.html">Overview</a>
</li>
        
<li>
  <a href="../analysis.html">Analysis</a>
</li>
        
      </ul>
        
<ul class="nav navbar-nav navbar-right">
<li>
   <a href="http://github.com/percylinhai/rvnpl-notes"> <span class="fa fa-github"></span> </a>
</li>
</ul>
        
      </div><!--/.nav-collapse -->
  </div><!--/.container -->
</div><!--/.navbar -->
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h1 id="RV-QNPL-analysis-for-simulated-genes">RV-QNPL analysis for simulated genes<a class="anchor-link" href="#RV-QNPL-analysis-for-simulated-genes">&#182;</a></h1>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Aim">Aim<a class="anchor-link" href="#Aim">&#182;</a></h2><p>In this notebook I show the workflow process to simulate family data with genotypes generated conditional on quantitative trait values using RarePedSim. And use RV-QNPL to analyze the simulated families.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Method-&amp;-Workflow-overview">Method &amp; Workflow overview<a class="anchor-link" href="#Method-&amp;-Workflow-overview">&#182;</a></h2>
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<ol>
<li><p>Simulate family data using RarePedSim assuming a linear mean shift disease model with the output of VCF file. (The parameters for RarePedSim are located in a configuration file generated by workflow below)</p>
</li>
<li><p>Then use "collapse" command from rvnpl to generate CHP regional markers for the genotypes in pedigrees</p>
</li>
<li><p>Finally, use "qnpl" command from rvnpl to analyze RV-QNPL score on the genotypes for CHP markers</p>
</li>
</ol>
<p>The default workflow can be executed by doing the following command:</p>

<pre><code>sos run analysis/QNPL_Simulation.ipynb simulate</code></pre>
<p>Or, we can simulate only a proportion of functional rare variants contribute to the disease (e.g. 50%):</p>

<pre><code>sos run analysis/QNPL_Simulation.ipynb simulate --name quant_Prop50 --proportion 0.5</code></pre>
<p>Or, we can simulate genotypes under the null by setting mean shift to 0:</p>

<pre><code>sos run analysis/QNPL_Simulation.ipynb simulate --name quant_null --mean_shift 0.0</code></pre>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h2 id="Input-Data">Input Data<a class="anchor-link" href="#Input-Data">&#182;</a></h2><ol>
<li>The PED file containing information of families</li>
<li>The SFS file containing the variant information for the simulated gene (6-column format: gene, chromosome, position, ref, alt, MAF, function score)*. </li>
</ol>
<p>*For the SFS files used here, the variant information was downloaded from gnomAD website. The non-Finnish European allele frequency is used for the MAF column. The value for the "function score" column is based on the functional annotations from gnomAD.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Global-Parameter-Setting">Global Parameter Setting<a class="anchor-link" href="#Global-Parameter-Setting">&#182;</a></h3>
</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[1]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="k">global</span><span class="p">]</span>
<span class="c1"># Disease model scenario</span>
<span class="kn">parameter:</span> <span class="kp">name</span> <span class="o">=</span> <span class="s1">&#39;quant_Prop100&#39;</span>
<span class="c1"># proportion of functional variants that contribute to the disease</span>
<span class="kn">parameter:</span> <span class="n">proportion</span> <span class="o">=</span> <span class="s1">&#39;None&#39;</span>
<span class="c1"># Mean shift value of detrimental rare variants</span>
<span class="kn">parameter:</span> <span class="n">mean_shift</span> <span class="o">=</span> <span class="mf">1.0</span>
<span class="c1"># model: LOGIT for qualitative traits or LNR for quantitative traits</span>
<span class="kn">parameter:</span> <span class="n">model</span> <span class="o">=</span> <span class="s1">&#39;LNR&#39;</span>
<span class="c1"># Path to the ped file (6-column PED format),quantitative trait values are standardized</span>
<span class="kn">parameter:</span> <span class="n">ped_file</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;data/100extend_quant.ped&#39;</span><span class="p">)</span>
<span class="c1"># Path to list of genes</span>
<span class="kn">parameter:</span> <span class="n">gene_list</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;data/genes.txt&#39;</span><span class="p">)</span>
<span class="c1"># the output directory for VCF file</span>
<span class="kn">parameter:</span> <span class="n">out_dir</span> <span class="o">=</span> <span class="n">path</span><span class="p">(</span><span class="s1">&#39;output&#39;</span><span class="p">)</span>

<span class="c1"># gene names</span>
<span class="n">genes</span> <span class="o">=</span> <span class="n">paths</span><span class="p">([</span><span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">gene_list</span><span class="si">:</span><span class="s1">d</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">x</span><span class="o">.</span><span class="n">strip</span><span class="p">()</span><span class="si">}</span><span class="s1">.sfs&#39;</span> <span class="k">for</span> <span class="n">x</span> <span class="ow">in</span> <span class="nb">open</span><span class="p">(</span><span class="n">gene_list</span><span class="p">)</span><span class="o">.</span><span class="n">readlines</span><span class="p">()])</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Configuration-file-for-disease-model">Configuration file for disease model<a class="anchor-link" href="#Configuration-file-for-disease-model">&#182;</a></h3><p>First, we will need to generate the configuration file for the disease model which will be used in the simulation software RarePedSim.</p>
<h4 id="Specify-various-simulation-scenarios-through-the-configuration-file">Specify various simulation scenarios through the configuration file<a class="anchor-link" href="#Specify-various-simulation-scenarios-through-the-configuration-file">&#182;</a></h4><p>The configuration file serves as the primary input for <code>RarePedSim</code> to generate simulated genotypes for families with given phenotype values. For quantitative traits, linear mean-shift ("LNR") model is used to model the shift effect of causal rare variants on the distribution of quantitative trait values. Furthermore, we can specify different simulation scenarios by giving different values for parameters in the configuration file. For example:</p>
<ol>
<li>when simulating genotypes under the null, we can specify "meanshift_rare_detrimental=0.0" to assume that detrimental rare variants have no impact on the distribution of quantitative trait value under study. That is, quantitative traits will be sampled from distribution $N(0, \sigma^2)$ for all genes regardless of whether or not rare variants present in the gene.</li>
<li>when simulating under the alternative, we used "meanshift_rare_detrimental=1.0" to assume carriers of detrimental rare variants are more likely to have larger trait value. That is, quantatitive traits will be sampled from distribution $N(\sigma M, \sigma^2)$ where $M$ is the number of rare variants in the gene.</li>
<li>when simulating with different proportion of rare variants being causal, we can specify a value for parameter "proportion_causal"(e.g. proportion_causal=0.5). In this case the number of rare variants contributing to the phenotype is $M' \sim Binom(M, p=0.5)$. </li>
</ol>
<p>For more details about the parameters in the configuration file, refer to <a href="https://github.com/statgenetics/rarepedsim/blob/master/doc/doc_RarePedSim.pdf">RarePedSim Documentation</a></p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">make_config</span><span class="p">:</span> <span class="kp">provides</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.conf&#39;</span><span class="p">]</span>
<span class="c1"># conf file contains the simulation specifications (either Mendelian or Complex, details in RarePedSim doc)</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.conf&#39;</span>
<span class="kn">report:</span> <span class="n">expand</span><span class="o">=</span><span class="kc">True</span><span class="p">,</span> <span class="n">output</span><span class="o">=</span><span class="n">_output</span>
    <span class="n">trait_type</span><span class="o">=</span><span class="n">Complex</span>
    <span class="p">[</span><span class="n">model</span><span class="p">]</span>
    <span class="n">model</span><span class="o">=</span><span class="p">{</span><span class="n">model</span><span class="p">}</span>
    <span class="p">[</span><span class="n">quality</span> <span class="n">control</span><span class="p">]</span>
    <span class="n">def_rare</span><span class="o">=</span><span class="mf">0.01</span>
    <span class="n">rare_only</span><span class="o">=</span><span class="kc">True</span>
    <span class="n">def_neutral</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mf">1E-5</span><span class="p">,</span> <span class="mf">1E-5</span><span class="p">)</span>
    <span class="n">def_protective</span><span class="o">=</span><span class="p">(</span><span class="o">-</span><span class="mi">1</span><span class="p">,</span> <span class="o">-</span><span class="mf">1E-5</span><span class="p">)</span>
    <span class="p">[</span><span class="n">phenotype</span> <span class="n">parameters</span><span class="p">]</span>
    <span class="n">baseline_effect</span><span class="o">=</span><span class="mf">0.01</span>
    <span class="n">moi</span><span class="o">=</span><span class="n">MAV</span>
    <span class="n">proportion_causal</span><span class="o">=</span><span class="p">{</span><span class="n">proportion</span><span class="p">}</span>
    <span class="p">[</span><span class="n">LOGIT</span> <span class="n">model</span><span class="p">]</span>
    <span class="n">OR_rare_detrimental</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">OR_rare_protective</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">ORmax_rare_detrimental</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">ORmin_rare_protective</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">OR_common_detrimental</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">OR_common_protective</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">[</span><span class="n">LNR</span> <span class="n">model</span><span class="p">]</span>
    <span class="n">meanshift_rare_detrimental</span><span class="o">=</span><span class="p">{</span><span class="n">mean_shift</span><span class="p">}</span>
    <span class="n">meanshift_rare_protective</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">meanshiftmax_rare_detrimental</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">meanshiftmax_rare_protective</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">meanshift_common_detrimental</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">meanshift_common_protective</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">[</span><span class="n">genotyping</span> <span class="n">artifact</span><span class="p">]</span>
    <span class="n">missing_low_maf</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">missing_sites</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">missing_calls</span><span class="o">=</span><span class="kc">None</span>
    <span class="n">error_calls</span><span class="o">=</span><span class="kc">None</span>
    <span class="p">[</span><span class="n">other</span><span class="p">]</span>
    <span class="n">max_vars</span><span class="o">=</span><span class="mi">2</span>
    <span class="n">ascertainment_qualitative</span><span class="o">=</span><span class="p">(</span><span class="mi">0</span><span class="p">,</span><span class="mi">0</span><span class="p">)</span>
    <span class="n">ascertainment_quantitative</span><span class="o">=</span><span class="p">((</span><span class="mi">0</span><span class="p">,</span><span class="o">~</span><span class="p">),(</span><span class="mi">0</span><span class="p">,</span><span class="o">~</span><span class="p">))</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generate-genotypes-for-given-families">Generate genotypes for given families<a class="anchor-link" href="#Generate-genotypes-for-given-families">&#182;</a></h3><p>Here, we use RarePedSim to generate genotypes for families with given quantitative traits based on user-specified disease model in the configuration file.</p>
<p>The output file is a VCF file and we need to tabix it before the next step.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">simulate_1</span> <span class="p">(</span><span class="n">rarepedsim</span><span class="p">)]</span>
<span class="kn">depends:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="kp">name</span><span class="si">}</span><span class="s1">.conf&#39;</span>
<span class="kn">input:</span> <span class="n">for_each</span> <span class="o">=</span> <span class="s1">&#39;genes&#39;</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">out_dir</span><span class="si">}</span><span class="s1">/</span><span class="si">{</span><span class="n">_genes</span><span class="si">:</span><span class="s1">bn</span><span class="si">}</span><span class="s1">.vcf.gz&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/rarepedsim&#39;</span><span class="p">,</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span>
    <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span><span class="o">.</span><span class="n">tbi</span> <span class="o">&amp;&amp;</span> <span class="n">mkdir</span> <span class="o">-</span><span class="n">p</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span>
    <span class="n">rarepedsim</span> <span class="n">generate</span> <span class="o">-</span><span class="n">s</span> <span class="err">$</span><span class="p">{</span><span class="n">_genes</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> <span class="o">-</span><span class="n">c</span> <span class="err">$</span><span class="p">{</span><span class="n">out_dir</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="kp">name</span><span class="p">}</span><span class="o">.</span><span class="n">conf</span> <span class="o">-</span><span class="n">p</span> <span class="err">$</span><span class="p">{</span><span class="n">ped_file</span><span class="p">:</span><span class="n">a</span><span class="p">}</span> <span class="o">--</span><span class="n">num_genes</span> <span class="mi">1</span> <span class="o">--</span><span class="n">num_reps</span> <span class="mi">1</span> <span class="o">-</span><span class="n">o</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span> <span class="o">--</span><span class="n">vcf</span> <span class="o">-</span><span class="n">b</span> <span class="o">-</span><span class="mi">1</span> \
    <span class="o">&amp;&amp;</span> <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="o">/</span><span class="err">$</span><span class="p">{</span><span class="n">_genes</span><span class="p">:</span><span class="n">bn</span><span class="p">}</span><span class="o">/</span><span class="n">rep1</span><span class="o">.</span><span class="n">vcf</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> <span class="o">&amp;&amp;</span> <span class="n">rm</span> <span class="o">-</span><span class="n">rf</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span>
    <span class="n">bgzip</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">n</span><span class="p">}</span> <span class="o">&amp;&amp;</span> <span class="n">tabix</span> <span class="o">-</span><span class="n">p</span> <span class="n">vcf</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Generate-CHP-regional-marker">Generate CHP regional marker<a class="anchor-link" href="#Generate-CHP-regional-marker">&#182;</a></h3><p>The next step is to use RV-NPL to generate CHP regional markers for the genotypes in families.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[&nbsp;]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">simulate_2</span> <span class="p">(</span><span class="n">CHP</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">nn</span><span class="si">}</span><span class="s1">/MERLIN/</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">bnn</span><span class="si">}</span><span class="s1">.CHP.ped&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/rvnpl&#39;</span><span class="p">,</span> <span class="n">environment</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HOME&#39;</span><span class="p">:</span> <span class="s1">&#39;/seqlink&#39;</span><span class="p">},</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">rvnpl</span> <span class="n">collapse</span> <span class="o">--</span><span class="n">fam</span> <span class="err">$</span><span class="p">{</span><span class="n">ped_file</span><span class="p">}</span> <span class="o">--</span><span class="n">vcf</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">}</span> <span class="o">--</span><span class="n">output</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span> <span class="o">--</span><span class="n">freq</span> <span class="n">EVSMAF</span> <span class="o">-</span><span class="n">c</span> <span class="mf">0.01</span> <span class="o">--</span><span class="n">rvhaplo</span> \
    <span class="o">&amp;&amp;</span> <span class="n">mv</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">nn</span><span class="p">}</span><span class="o">.</span><span class="n">chr</span><span class="o">*.</span><span class="n">ped</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">}</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Perform-RV-QNPL-analysis">Perform RV-QNPL analysis<a class="anchor-link" href="#Perform-RV-QNPL-analysis">&#182;</a></h3><p>Finally, we use RV-QNPL to analyze the CHP genotypes to get the significance of allele-sharing on rare variants in the families.</p>

</div>
</div>
</div>
<div class="cell border-box-sizing code_cell rendered">
<div class="input">
<div class="prompt input_prompt">In&nbsp;[2]:</div>
<div class="inner_cell">
    <div class="input_area">
<div class=" highlight hl-sos"><pre><span></span><span class="p">[</span><span class="n">simulate_3</span> <span class="p">(</span><span class="n">rvnpl</span><span class="p">)]</span>
<span class="kn">output:</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_input</span><span class="si">:</span><span class="s1">dd</span><span class="si">}</span><span class="s1">/pvalue.txt&#39;</span>
<span class="kn">bash:</span> <span class="n">container</span> <span class="o">=</span> <span class="s1">&#39;statisticalgenetics/rvnpl&#39;</span><span class="p">,</span> <span class="n">environment</span><span class="o">=</span><span class="p">{</span><span class="s1">&#39;HOME&#39;</span><span class="p">:</span> <span class="s1">&#39;/seqlink&#39;</span><span class="p">},</span> <span class="n">expand</span> <span class="o">=</span> <span class="s1">&#39;${ }&#39;</span><span class="p">,</span> <span class="n">stderr</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stderr&#39;</span><span class="p">,</span> <span class="n">stdout</span> <span class="o">=</span> <span class="sa">f</span><span class="s1">&#39;</span><span class="si">{</span><span class="n">_output</span><span class="si">:</span><span class="s1">n</span><span class="si">}</span><span class="s1">.stdout&#39;</span>
    <span class="n">rvnpl</span> <span class="n">qnpl</span> <span class="o">--</span><span class="n">path</span> <span class="err">$</span><span class="p">{</span><span class="n">_input</span><span class="p">:</span><span class="n">dd</span><span class="p">}</span> <span class="o">--</span><span class="n">output</span> <span class="err">$</span><span class="p">{</span><span class="n">_output</span><span class="p">:</span><span class="n">d</span><span class="p">}</span> <span class="o">--</span><span class="n">exact</span> <span class="o">--</span><span class="n">rvibd</span> <span class="o">--</span><span class="n">n_jobs</span> <span class="mi">8</span> <span class="o">-</span><span class="n">c</span> <span class="mf">0.001</span> <span class="o">--</span><span class="n">lower_cut</span> <span class="mf">1E-8</span> <span class="o">--</span><span class="n">rep</span> <span class="mi">2000000</span>
</pre></div>

    </div>
</div>
</div>

</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<h3 id="Results">Results<a class="anchor-link" href="#Results">&#182;</a></h3><Add Result> 
</div>
</div>
</div>
<div class="cell border-box-sizing text_cell rendered"><div class="prompt input_prompt">
</div><div class="inner_cell">
<div class="text_cell_render border-box-sizing rendered_html">
<p>The p-values for two QNPL scores are presented in file pvalue.txt in the corresponding gene folder under the output directory.</p>

</div>
</div>
</div>
<hr>
&copy 2019 Linhai Zhao at Center for Statistical Genetics, Baylor College of Medicine
<p><small>Exported from <a href="http://github.com/percylinhai/rvnpl-notes/blob/554cb6d25f62da1743abc50190bbc596d0bd36f0/analysis/QNPL_Simulation.ipynb"><code>analysis/QNPL_Simulation.ipynb</code></a> committed by Gao Wang on Wed Apr 29 18:33:00 2020 <a href="http://github.com/percylinhai/rvnpl-notes/commit/554cb6d25f62da1743abc50190bbc596d0bd36f0">revision 4, 554cb6d</a> <a href="https://stephenslab.github.io/ipynb-website/notes.html#Note-about-commit-ids"><span class="fa fa-question-circle"></span></a></small></p>
</div>
</div>
</body>
</html>
