{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# RV-NPL analysis for simulated genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Aim\n",
    "\n",
    "In this notebook I show the workflow process to simulate family data with genotypes generated conditional on affection status using RarePedSim. And use RV-NPL to analyze the simulated families."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Method & Workflow overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "markdown"
   },
   "source": [
    "1. Simulate family data using RarePedSim with the output of VCF file. (The parameters for RarePedSim are located in a configuration file generated by workflow below)\n",
    "2. Then use \"collapse\" command from rvnpl to generate CHP regional markers for the genotypes in pedigrees\n",
    "3. Finally, use \"npl\" command from rvnpl to analyze RV-NPL score on the genotypes for CHP markers\n",
    "\n",
    "The default workflow can be executed by doing the following command:\n",
    "\n",
    "```\n",
    "sos run analysis/NPL_Simulation.ipynb simulate\n",
    "```\n",
    "Or, we can simulate a proportion of functional rare variants contribute to the disease:\n",
    "```\n",
    "sos run analysis/NPL_Simulation.ipynb simulate --name Prop50 --proportion 0.5\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input Data\n",
    "1. The PED file containing information of families\n",
    "2. The SFS file containing the variant information for the simulated gene (6-column format: gene, chromosome, position, ref, alt, MAF, function score)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Global Parameter Setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Disease model scenario\n",
    "parameter: name = 'Prop100'\n",
    "# proportion of functional variants that contribute to the disease\n",
    "parameter: proportion = 'None'\n",
    "# Odds Ratio\n",
    "parameter: OR = 5.0\n",
    "# model: LOGIT for qualitative traits or LNR for quantitative traits\n",
    "parameter: model = 'LOGIT'\n",
    "# Path to the ped file (6-column PED format)\n",
    "parameter: ped_file = path('data/100extend01.ped')\n",
    "# Path to list of genes\n",
    "parameter: gene_list = path('data/genes.txt')\n",
    "# the output directory for VCF file\n",
    "parameter: out_dir = path('output')\n",
    "\n",
    "# gene names\n",
    "genes = paths([f'{gene_list:d}/{x.strip()}.sfs' for x in open(gene_list).readlines()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Configuration file for disease model\n",
    "First, we will need to generate the configuration file for the disease model which will be used in the simulation software RarePedSim."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[make_config: provides = f'{out_dir}/{name}.conf']\n",
    "# conf file contains the simulation specifications (either Mendelian or Complex, details in RarePedSim doc)\n",
    "output: f'{out_dir}/{name}.conf'\n",
    "report: expand=True, output=_output\n",
    "    trait_type=Complex\n",
    "    [model]\n",
    "    model={model}\n",
    "    [quality control]\n",
    "    def_rare=0.01\n",
    "    rare_only=True\n",
    "    def_neutral=(-1E-5, 1E-5)\n",
    "    def_protective=(-1, -1E-5)\n",
    "    [phenotype parameters]\n",
    "    baseline_effect=0.01\n",
    "    moi=MAV\n",
    "    proportion_causal={proportion}\n",
    "    [LOGIT model]\n",
    "    OR_rare_detrimental={OR}\n",
    "    OR_rare_protective=None\n",
    "    ORmax_rare_detrimental=None\n",
    "    ORmin_rare_protective=None\n",
    "    OR_common_detrimental=None\n",
    "    OR_common_protective=None\n",
    "    [LNR model]\n",
    "    meanshift_rare_detrimental=0.0\n",
    "    meanshift_rare_protective=None\n",
    "    meanshiftmax_rare_detrimental=None\n",
    "    meanshiftmax_rare_protective=None\n",
    "    meanshift_common_detrimental=None\n",
    "    meanshift_common_protective=None\n",
    "    [genotyping artifact]\n",
    "    missing_low_maf=None\n",
    "    missing_sites=None\n",
    "    missing_calls=None\n",
    "    error_calls=None\n",
    "    [other]\n",
    "    max_vars=2\n",
    "    ascertainment_qualitative=(0,0)\n",
    "    ascertainment_quantitative=((0,~),(0,~))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Generate genotypes for given families\n",
    "Here, we use RarePedSim to generate genotypes for families with given affection status based on user-specified disease model in the configuration file. \n",
    "\n",
    "The output file is a VCF file and we need to tabix it before the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[simulate_1 (rarepedsim)]\n",
    "depends: f'{out_dir}/{name}.conf'\n",
    "input: for_each = 'genes'\n",
    "output: f'{out_dir}/{_genes:bn}.vcf.gz'\n",
    "bash: container = 'statisticalgenetics/rvnpl', expand = '${ }'\n",
    "    rm -rf ${_output:nn} ${_output} ${_output}.tbi && mkdir -p ${_output:nn}\n",
    "    rarepedsim generate -s ${_genes:a} -c ${out_dir}/${name}.conf -p ${ped_file:a} --num_genes 1 --num_reps 1 -o ${_output:nn} --vcf -b -1 \\\n",
    "    && mv ${_output:nn}/${_genes:bn}/rep1.vcf ${_output:n} && rm -rf ${_output:nn}\n",
    "    bgzip ${_output:n} && tabix -p vcf ${_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Generate CHP regional marker\n",
    "The next step is to use RV-NPL to generate CHP regional markers for the genotypes in families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[simulate_2 (CHP)]\n",
    "output: f'{_input:nn}/MERLIN/{_input:bnn}.CHP.ped'\n",
    "bash: container = 'statisticalgenetics/rvnpl', expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    rvnpl collapse --fam ${ped_file} --vcf ${_input} --output ${_input:nn} --freq EVSMAF -c 0.01 --rvhaplo \\\n",
    "    && mv ${_output:nn}.chr*.ped ${_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Perform RV-NPL analysis\n",
    "Finally, we use RV-NPL to analyze the CHP genotypes to get the significance of allele-sharing on rare variants in the families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[simulate_3 (rvnpl)]\n",
    "output: f'{_input:dd}/pvalue.txt'\n",
    "bash: container = 'statisticalgenetics/rvnpl', expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    rvnpl npl --path ${_input:dd} --output ${_output:d} --exact --info_only --perfect --sall --rvibd --n_jobs 8 -c 0.001 --lower_cut 1E-8 --rep 2000000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Results\n",
    "\n",
    "<Add Result> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "The p-values for two NPL scores are presented in file pvalue.txt in the corresponding gene folder under the output directory."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ],
    [
     "markdown",
     "markdown",
     "markdown",
     "",
     ""
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0
   },
   "version": "0.20.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
