{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# Evaluation of RV-QNPL Through Simulation Studies"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Aim\n",
    "\n",
    "The aim of this notebook is to display a workflow process to 1) simulate genotype data for families, conditional on their quantitative trait (QT) values using RarePedSim and 2) analysis of the simulated data using RV-QNPL to evaluate type I error and power."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Method & Workflow overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "markdown"
   },
   "source": [
    "1. [`RarePedSim`](http://www.bioinformatics.org/simped/rare/) is used to simulate genotypes for family data using a linear mean shift model with the output of VCF file. (The parameters for RarePedSim are located in a configuration file generated by workflow shown below)\n",
    "\n",
    "2. To generate the collapsed haplotype pattern (CHP) regional markers for pedigrees the \"collapse\" command from `rvnpl` is used.\n",
    "\n",
    "3. The \"qnpl\" command from `rvnpl` is used to analyze the QT and CHP regional markers.\n",
    "\n",
    "The default workflow can be executed by doing the following command:\n",
    "\n",
    "```\n",
    "sos run analysis/QNPL_Simulation.ipynb\n",
    "```\n",
    "\n",
    "Alternatively, simulations can also be performed so that only a proportion of functional rare variants, e.g., 50%, contribute to disease etiology:\n",
    "\n",
    "```\n",
    "sos run analysis/QNPL_Simulation.ipynb --name quant_Prop50 --proportion 0.5\n",
    "```\n",
    "\n",
    "It is also possible to simulate genotypes under the null (no linkage) by setting the mean shift to 0:\n",
    "\n",
    "```\n",
    "sos run analysis/QNPL_Simulation.ipynb --name quant_null --mean_shift 0.0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input Data\n",
    "1. The PED file contains the pedigree structures as well as QT values.\n",
    "2. The SFS file containing the variant information for each simulated gene (6-column format: gene, chromosome, position, ref, alt, MAF, function score).\n",
    "\n",
    "For the SFS files used here, the variant information was downloaded from gnomAD website. Non-Finnish European (NFE) allele frequencies are used in the minor allele frequency (MAF) column. The value for the \"function score\" column is based on the functional annotations from gnomAD which determines the type of variant, i.e. synonymous, missense. Each SFS file contains variants from each simulated gene respectively. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Global Parameter Setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Disease model scenario\n",
    "parameter: name = 'quant_Prop100'\n",
    "# proportion of functional variants that contribute to the disease\n",
    "parameter: proportion = 'None'\n",
    "# Mean shift value of detrimental rare variants\n",
    "parameter: mean_shift = 1.0\n",
    "# model: LOGIT for qualitative traits or LNR for quantitative traits\n",
    "parameter: model = 'LNR'\n",
    "# Path to the ped file (6-column PED format),quantitative trait values are standardized\n",
    "parameter: ped_file = path('data/100extend_quant.ped')\n",
    "# Path to list of genes\n",
    "parameter: gene_list = path('data/genes.txt')\n",
    "# the output directory for VCF file\n",
    "parameter: out_dir = path('output')\n",
    "\n",
    "# gene names\n",
    "genes = paths([f'{gene_list:d}/{x.strip()}.sfs' for x in open(gene_list).readlines()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Configuration file for disease model\n",
    "First, we will need to generate the configuration file for the disease model which will be used in the simulation software `RarePedSim`.\n",
    "\n",
    "#### Specify various simulation scenarios through the configuration file\n",
    "\n",
    "The configuration file serves as the primary input for `RarePedSim` to generate simulated genotypes for families with QT values. For QTs, the linear mean-shift (\"LNR\") model is used to model the effect of causal rare variants on the distribution of QT values. The input PED file contains a column of QT values. These QT values were generated by  sampling from a $N(2,1)$ distribution for a subset of family members and from a $N(0,1)$ distribution for the remaining family members, then standardized (mean = 0, standard deviation = 1) and saved in the last column of the PED file. Given the QT values in the pedigree(s) and MAF data obtained from gnomAD, genotypes can be simulated. For example:\n",
    "\n",
    "1. When genotypes are simulated under the null, rare variants are not associated with QT values, i.e., `meanshift_rare_detrimental=0.0`. Genotype are generated using MAF obtained from gnomAD assuming Hardy-Weinberg equilibrium and Mendelian segregation, regardless of QT values.\n",
    "\n",
    "2. When genotypes are simulated under the alternative, rare variants are assumed to cause an increase in QT values, i.e., `meanshift_rare_detrimental` is set to >0. For this example, we set `meanshift_rare_detrimental=1` so that rare variant within a gene region increase with the QT value. The distribution of rare variants is $N(\\sigma M, \\sigma^2)$ where $M$ is total count of detrimental rare variants within a gene. Conditional on pre-specified QT values, the distribution of $M$, denoted as $p(M)$ are estimated and the genotypes are assigned based on Mendelian segregation. For formal derivation of this model please refer to Section 3.3 of [RarePedSim Documentation](https://github.com/statgenetics/rarepedsim/blob/master/doc/doc_RarePedSim.pdf).\n",
    "3. When it is desired to simulate different proportions of rare variants being causal (rare variants that increase QT values) `proportion_causal` can be set be set <1, e.g., `proportion_causal=0.5`.  When all variants are causal we set `proportion_causal=1`.\n",
    "\n",
    "Please refer to the Appendix of [RarePedSim Documentation](https://github.com/statgenetics/rarepedsim/blob/master/doc/doc_RarePedSim.pdf) for more details on other parameters used in the configuration file."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[make_config: provides = f'{out_dir}/{name}.conf']\n",
    "# conf file contains the simulation specifications (either Mendelian or Complex, details in RarePedSim doc)\n",
    "output: f'{out_dir}/{name}.conf'\n",
    "report: expand=True, output=_output\n",
    "    trait_type=Complex\n",
    "    [model]\n",
    "    model={model}\n",
    "    [quality control]\n",
    "    def_rare=0.01\n",
    "    rare_only=True\n",
    "    def_neutral=(-1E-5, 1E-5)\n",
    "    def_protective=(-1, -1E-5)\n",
    "    [phenotype parameters]\n",
    "    baseline_effect=0.01\n",
    "    moi=MAV\n",
    "    proportion_causal={proportion}\n",
    "    [LOGIT model]\n",
    "    OR_rare_detrimental=None\n",
    "    OR_rare_protective=None\n",
    "    ORmax_rare_detrimental=None\n",
    "    ORmin_rare_protective=None\n",
    "    OR_common_detrimental=None\n",
    "    OR_common_protective=None\n",
    "    [LNR model]\n",
    "    meanshift_rare_detrimental={mean_shift}\n",
    "    meanshift_rare_protective=None\n",
    "    meanshiftmax_rare_detrimental=None\n",
    "    meanshiftmax_rare_protective=None\n",
    "    meanshift_common_detrimental=None\n",
    "    meanshift_common_protective=None\n",
    "    [genotyping artifact]\n",
    "    missing_low_maf=None\n",
    "    missing_sites=None\n",
    "    missing_calls=None\n",
    "    error_calls=None\n",
    "    [other]\n",
    "    max_vars=2\n",
    "    ascertainment_qualitative=(0,0)\n",
    "    ascertainment_quantitative=((0,~),(0,~))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Generating Genotypes for Families\n",
    "\n",
    "RarePedSim is used to generate genotypes for families with QT data based on user-specified model in the configuration file.\n",
    "\n",
    "The output is a VCF file and `tabix` tool is used to build index in preparation for input to the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[default_1 (RarePedSim simulation of pedigree genotype data)]\n",
    "depends: f'{out_dir}/{name}.conf'\n",
    "input: for_each = 'genes'\n",
    "output: f'{out_dir}/{_genes:bn}.vcf.gz'\n",
    "bash: container = 'statisticalgenetics/rarepedsim', expand = '${ }'\n",
    "    rm -rf ${_output:nn} ${_output} ${_output}.tbi && mkdir -p ${_output:nn}\n",
    "    rarepedsim generate -s ${_genes:a} -c ${out_dir}/${name}.conf -p ${ped_file:a} --num_genes 1 --num_reps 1 -o ${_output:nn} --vcf -b -1 \\\n",
    "    && mv ${_output:nn}/${_genes:bn}/rep1.vcf ${_output:n} && rm -rf ${_output:nn}\n",
    "    bgzip ${_output:n} && tabix -p vcf ${_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Construction of CHP Regional Markers using RV-NPL\n",
    "\n",
    "RV-NPL determines the CHP regional markers using genotypes from the families. Variants from either simulation studies or from sequence data can be used as input data. Here we are using simulated variants from step `default_1`. Rare variants with MAF below 0.01 are analyzed to construct the CHP regional markers by setting the parameter `-c 0.01`. The MAF information column is specified at `--freq EVSMAF`, since only variants with a MAF<0.01 will be analyzed "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[default_2 (Computing CHP markers)]\n",
    "output: f'{_input:nn}/MERLIN/{_input:bnn}.CHP.ped'\n",
    "bash: container = 'statisticalgenetics/rvnpl', environment={'HOME': '/seqlink'}, expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    rvnpl collapse --fam ${ped_file} --vcf ${_input} --output ${_input:nn} --freq EVSMAF -c 0.01 --rvhaplo \\\n",
    "    && mv ${_output:nn}.chr*.ped ${_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Performing RV-QNPL Analysis on QT Values and CHP Regional Markers\n",
    "RV-QNPL is used to analyze the CHP genotypes and QT values.  Unlike NPL, for RV-NPL only sharing of the minor allele is evaluated. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[default_3 (NPL analysis of rare variants for quantitative traits)]\n",
    "output: f'{_input:dd}/pvalue.txt'\n",
    "bash: container = 'statisticalgenetics/rvnpl', environment={'HOME': '/seqlink'}, expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    rvnpl qnpl --path ${_input:dd} --output ${_output:d} --exact --rvibd --n_jobs 8 -c 0.001 --lower_cut 1E-8 --rep 2000000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Results\n",
    "\n",
    "<Add Result> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "The p-values for two QNPL scores are presented in file pvalue.txt in the corresponding gene folder under the output directory."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ],
    [
     "markdown",
     "markdown",
     "markdown",
     "",
     ""
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0
   },
   "version": "0.21.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
