{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "# RV-QNPL analysis for simulated genes"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Aim\n",
    "\n",
    "In this notebook I show the workflow process to simulate family data with genotypes generated conditional on quantitative trait values using RarePedSim. And use RV-QNPL to analyze the simulated families."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Method & Workflow overview"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "markdown"
   },
   "source": [
    "1. Simulate family data using RarePedSim assuming a linear mean shift disease model with the output of VCF file. (The parameters for RarePedSim are located in a configuration file generated by workflow below)\n",
    "\n",
    "2. Then use \"collapse\" command from rvnpl to generate CHP regional markers for the genotypes in pedigrees\n",
    "\n",
    "3. Finally, use \"qnpl\" command from rvnpl to analyze RV-QNPL score on the genotypes for CHP markers\n",
    "\n",
    "The default workflow can be executed by doing the following command:\n",
    "\n",
    "```\n",
    "sos run analysis/QNPL_Simulation.ipynb simulate\n",
    "```\n",
    "\n",
    "Or, we can simulate only a proportion of functional rare variants contribute to the disease (e.g. 50%):\n",
    "```\n",
    "sos run analysis/QNPL_Simulation.ipynb simulate --name quant_Prop50 --proportion 0.5\n",
    "```\n",
    "\n",
    "Or, we can simulate genotypes under the null by setting mean shift to 0:\n",
    "```\n",
    "sos run analysis/QNPL_Simulation.ipynb simulate --name quant_null --mean_shift 0.0\n",
    "```"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "## Input Data\n",
    "1. The PED file containing information of families\n",
    "2. The SFS file containing the variant information for the simulated gene (6-column format: gene, chromosome, position, ref, alt, MAF, function score)*. \n",
    "\n",
    "*For the SFS files used here, the variant information was downloaded from gnomAD website. The non-Finnish European allele frequency is used for the MAF column. The value for the \"function score\" column is based on the functional annotations from gnomAD. "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Global Parameter Setting"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[global]\n",
    "# Disease model scenario\n",
    "parameter: name = 'quant_Prop100'\n",
    "# proportion of functional variants that contribute to the disease\n",
    "parameter: proportion = 'None'\n",
    "# Mean shift value of detrimental rare variants\n",
    "parameter: mean_shift = 1.0\n",
    "# model: LOGIT for qualitative traits or LNR for quantitative traits\n",
    "parameter: model = 'LNR'\n",
    "# Path to the ped file (6-column PED format),quantitative trait values are standardized\n",
    "parameter: ped_file = path('data/100extend_quant.ped')\n",
    "# Path to list of genes\n",
    "parameter: gene_list = path('data/genes.txt')\n",
    "# the output directory for VCF file\n",
    "parameter: out_dir = path('output')\n",
    "\n",
    "# gene names\n",
    "genes = paths([f'{gene_list:d}/{x.strip()}.sfs' for x in open(gene_list).readlines()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Configuration file for disease model\n",
    "First, we will need to generate the configuration file for the disease model which will be used in the simulation software RarePedSim.\n",
    "\n",
    "#### Specify various simulation scenarios through the configuration file\n",
    "The configuration file serves as the primary input for RarePedSim to generate simulated genotypes for families with given phenotype values. For quantitative traits, linear mean-shift (\"LNR\") model is used to model the shift effect of causal rare variants on the distribution of quantitative trait values. Furthermore, we can specify different simulation scenarios by giving different values for parameters in the configuration file. For example:\n",
    "\n",
    "    a. when simulating genotypes under the null, we can specify \"meanshift_rare_detrimental=0.0\" to assume that detrimental rare variants have no impact on the distribution of quantitative trait value under study; \n",
    "    b. when simulating under the alternative, we used \"meanshift_rare_detrimental=1.0\" to assume carriers of detrimental rare variants are more likely to have larger trait value;\n",
    "    c. when simulating with different proportion of rare variants being causal, we can specify a value for parameter \"proportion_causal\"(e.g. proportion_causal=0.5);\n",
    "\n",
    "For more details about the parameters in the configuration file, refer to [RarePedSim Documentation](https://github.com/statgenetics/rarepedsim/blob/master/doc/doc_RarePedSim.pdf)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[make_config: provides = f'{out_dir}/{name}.conf']\n",
    "# conf file contains the simulation specifications (either Mendelian or Complex, details in RarePedSim doc)\n",
    "output: f'{out_dir}/{name}.conf'\n",
    "report: expand=True, output=_output\n",
    "    trait_type=Complex\n",
    "    [model]\n",
    "    model={model}\n",
    "    [quality control]\n",
    "    def_rare=0.01\n",
    "    rare_only=True\n",
    "    def_neutral=(-1E-5, 1E-5)\n",
    "    def_protective=(-1, -1E-5)\n",
    "    [phenotype parameters]\n",
    "    baseline_effect=0.01\n",
    "    moi=MAV\n",
    "    proportion_causal={proportion}\n",
    "    [LOGIT model]\n",
    "    OR_rare_detrimental=None\n",
    "    OR_rare_protective=None\n",
    "    ORmax_rare_detrimental=None\n",
    "    ORmin_rare_protective=None\n",
    "    OR_common_detrimental=None\n",
    "    OR_common_protective=None\n",
    "    [LNR model]\n",
    "    meanshift_rare_detrimental={mean_shift}\n",
    "    meanshift_rare_protective=None\n",
    "    meanshiftmax_rare_detrimental=None\n",
    "    meanshiftmax_rare_protective=None\n",
    "    meanshift_common_detrimental=None\n",
    "    meanshift_common_protective=None\n",
    "    [genotyping artifact]\n",
    "    missing_low_maf=None\n",
    "    missing_sites=None\n",
    "    missing_calls=None\n",
    "    error_calls=None\n",
    "    [other]\n",
    "    max_vars=2\n",
    "    ascertainment_qualitative=(0,0)\n",
    "    ascertainment_quantitative=((0,~),(0,~))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Generate genotypes for given families\n",
    "Here, we use RarePedSim to generate genotypes for families with given quantitative traits based on user-specified disease model in the configuration file. \n",
    "\n",
    "The output file is a VCF file and we need to tabix it before the next step."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[simulate_1 (rarepedsim)]\n",
    "depends: f'{out_dir}/{name}.conf'\n",
    "input: for_each = 'genes'\n",
    "output: f'{out_dir}/{_genes:bn}.vcf.gz'\n",
    "bash: container = 'statisticalgenetics/rarepedsim', expand = '${ }'\n",
    "    rm -rf ${_output:nn} ${_output} ${_output}.tbi && mkdir -p ${_output:nn}\n",
    "    rarepedsim generate -s ${_genes:a} -c ${out_dir}/${name}.conf -p ${ped_file:a} --num_genes 1 --num_reps 1 -o ${_output:nn} --vcf -b -1 \\\n",
    "    && mv ${_output:nn}/${_genes:bn}/rep1.vcf ${_output:n} && rm -rf ${_output:nn}\n",
    "    bgzip ${_output:n} && tabix -p vcf ${_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Generate CHP regional marker\n",
    "The next step is to use RV-NPL to generate CHP regional markers for the genotypes in families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[simulate_2 (CHP)]\n",
    "output: f'{_input:nn}/MERLIN/{_input:bnn}.CHP.ped'\n",
    "bash: container = 'statisticalgenetics/rvnpl', environment={'HOME': '/seqlink'}, expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    rvnpl collapse --fam ${ped_file} --vcf ${_input} --output ${_input:nn} --freq EVSMAF -c 0.01 --rvhaplo \\\n",
    "    && mv ${_output:nn}.chr*.ped ${_output}"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Perform RV-QNPL analysis\n",
    "Finally, we use RV-QNPL to analyze the CHP genotypes to get the significance of allele-sharing on rare variants in the families."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {
    "kernel": "SoS"
   },
   "outputs": [],
   "source": [
    "[simulate_3 (rvnpl)]\n",
    "output: f'{_input:dd}/pvalue.txt'\n",
    "bash: container = 'statisticalgenetics/rvnpl', environment={'HOME': '/seqlink'}, expand = '${ }', stderr = f'{_output:n}.stderr', stdout = f'{_output:n}.stdout'\n",
    "    rvnpl qnpl --path ${_input:dd} --output ${_output:d} --exact --rvibd --n_jobs 8 -c 0.001 --lower_cut 1E-8 --rep 2000000"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "### Results\n",
    "\n",
    "<Add Result> "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {
    "kernel": "SoS"
   },
   "source": [
    "The p-values for two QNPL scores are presented in file pvalue.txt in the corresponding gene folder under the output directory."
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "SoS",
   "language": "sos",
   "name": "sos"
  },
  "language_info": {
   "codemirror_mode": "sos",
   "file_extension": ".sos",
   "mimetype": "text/x-sos",
   "name": "sos",
   "nbconvert_exporter": "sos_notebook.converter.SoS_Exporter",
   "pygments_lexer": "sos"
  },
  "sos": {
   "kernels": [
    [
     "SoS",
     "sos",
     "",
     "",
     "sos"
    ],
    [
     "markdown",
     "markdown",
     "markdown",
     "",
     ""
    ]
   ],
   "panel": {
    "displayed": true,
    "height": 0
   },
   "version": "0.21.7"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
